File: Linked List/0021. Merge Two Sorted Lists/0021.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
	ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
/*
        if (l1 == NULL && l2 == NULL) {
            return NULL;
        }

        else if (l1 == NULL || l2 == NULL) {
            if (l1 == NULL) return l2;
            else if (l2 == NULL) return l1;
        }

        ListNode* head = NULL;

        if (l1->val <= l2->val) {
            head = l1;
            l1 = l1->next;
        }
        else {
            head = l2;
            l2 = l2->next;
        }

        ListNode* curr = head;
        while (l1 != NULL && l2 != NULL) {
            if (l1->val <= l2->val) {
                curr->next = l1;
                l1 = l1->next;
            }
            else {
                curr->next = l2;
                l2 = l2->next;
            }
            curr = curr->next;
        }

        if (l1 == NULL) {
            curr->next = l2;
        }
        else if (l2 == NULL){
            curr->next = l1;
        }

        return head;
*/

    if (l1 == NULL && l2 == NULL) {
        return NULL;
    }

    if (l1 == NULL) {
        return l2;
    }
    else if (l2 == NULL) {
        return l1;
    }

    if (l1->val <= l2->val) {
        l1->next = mergeTwoLists(l1->next, l2);
        return l1;
    }

    else {
        l2->next = mergeTwoLists(l1, l2->next);
        return l2;
    }

    }
};

File: Linked List/0021. Merge Two Sorted Lists/0021.py
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:

        # create a new node in case l1 and l2 do not exist, then we must just return an empty thing. Tail is simply a pointer to iterate
        dummy = ListNode()
        tail = dummy

        while list1 and list2:
            if list1.val > list2.val:
                tail.next = list1
                list1 = list1.next
            else:
                tail.next = list2
                list2 = list2.next
            # update the pointer (what tail points to, not what tail actually is)
            tail = tail.next
        
        # if one of the l1 or l2 still has values in it, can just smack our newlist to the start of the remaining ones (just because it is sorted)
        if list1:
            tail.next = list1
        elif list2:
            tail.next = list2
        
        # because dummy.next would be nullptr if not changed with tail
        return dummy.next
